libraries/cms/application/cms.php

```diff
@@ -114,103 +114,15 @@ public function __construct(JInput $input = null, Registry $config = null, JAppl
 		}
 
 		// Enable sessions by default.
-		if (is_null($this->config->get('session')))
+		if (is_null($this->getContainer()->get('config')->get('session')))
```

Comment:
> The component MUST NOT go through the application to get the container or the application configuration.

How the container will be passed at the end to the component is up to the implementation, was probably a bad example in my comment. At the end the application does run the code to execute the component. If it passes only the container to run the component, that's fine.

> Moreover, it should be possible for me to tell the application to use a different database connection or session object than the one we normally use in Joomla!.

On which part of the application life cycle you want to do that? There is nothing which prevents you to do that? An application knows how to load its database per default because it needs one. If you want to change it afterwards then you still can override the database in the container.

> Now we have to do that with every app instance which is far worse.

I don't understand that. What you mean with every app instance? You have only one app during a request.